cmd/medsky/bgs/bgs.go

```diff
@@ -0,0 +1,1440 @@
+
+	// NextCrawlers gets forwarded POST /xrpc/com.atproto.sync.requestCrawl
+	NextCrawlers []*url.URL
+
+	ApplyPDSClientSettings func(c *xrpc.Client)
+	InductionTraceLog      *slog.Logger
+}
+
+func DefaultBGSConfig() *BGSConfig {
+	return &BGSConfig{
+		SSL:               true,
+		DefaultRepoLimit:  100,
+		ConcurrencyPerPDS: 100,
+		MaxQueuePerPDS:    1_000,
+	}
+}
+
+func NewBGS(db *gorm.DB, repoman *repomgr.RepoManager, evtman *events.EventManager, didd identity.Directory, config *BGSConfig) (*BGS, error) {
+
+	if config == nil {
+		config = DefaultBGSConfig()
+	}
+	err := db.AutoMigrate(User{})
+	if err != nil {
+		panic(err)
+	}
+	err = db.AutoMigrate(UserPreviousState{})
+	if err != nil {
+		panic(err)
+	}
+	err = db.AutoMigrate(AuthToken{})
+	if err != nil {
+		panic(err)
+	}
+	err = db.AutoMigrate(models.PDS{})
+	if err != nil {
+		panic(err)
+	}
+	err = db.AutoMigrate(models.DomainBan{})
+	if err != nil {
+		panic(err)
+	}
+
+	uc, _ := lru.New[string, *User](1_000_000)
+
+	bgs := &BGS{
+		db: db,
+
+		repoman: repoman,
+		events:  evtman,
+		didd:    didd,
+		ssl:     config.SSL,
+
+		consumersLk: sync.RWMutex{},
+		consumers:   make(map[uint64]*SocketConsumer),
+
+		userCache: uc,
+
+		log: slog.Default().With("system", "bgs"),
+
+		config: *config,
+
+		inductionTraceLog: config.InductionTraceLog,
+	}
+
+	slOpts := DefaultSlurperOptions()
+	slOpts.SSL = config.SSL
+	slOpts.DefaultRepoLimit = config.DefaultRepoLimit
+	slOpts.ConcurrencyPerPDS = config.ConcurrencyPerPDS
+	slOpts.MaxQueuePerPDS = config.MaxQueuePerPDS
+	slOpts.Logger = bgs.log
+	s, err := NewSlurper(db, bgs.handleFedEvent, slOpts)
+	if err != nil {
+		return nil, err
+	}
+
+	bgs.slurper = s
+
+	if err := bgs.slurper.RestartAll(); err != nil {
+		return nil, err
+	}
+
+	bgs.nextCrawlers = config.NextCrawlers
+	bgs.httpClient.Timeout = time.Second * 5
+
+	return bgs, nil
+}
+
+func (bgs *BGS) StartMetrics(listen string) error {
+	http.Handle("/metrics", promhttp.Handler())
+	return http.ListenAndServe(listen, nil)
+}
+
+func (bgs *BGS) Start(addr string) error {
+	var lc net.ListenConfig
+	ctx, cancel := context.WithTimeout(context.Background(), serverListenerBootTimeout)
+	defer cancel()
+
+	li, err := lc.Listen(ctx, "tcp", addr)
+	if err != nil {
+		return err
+	}
+	return bgs.StartWithListener(li)
+}
+
+func (bgs *BGS) StartWithListener(listen net.Listener) error {
+	e := echo.New()
+	e.HideBanner = true
+
+	e.Use(middleware.CORSWithConfig(middleware.CORSConfig{
+		AllowOrigins: []string{"*"},
+		AllowHeaders: []string{echo.HeaderOrigin, echo.HeaderContentType, echo.HeaderAccept, echo.HeaderAuthorization},
+	}))
+
+	if !bgs.ssl {
+		e.Use(middleware.LoggerWithConfig(middleware.LoggerConfig{
+			Format: "method=${method}, uri=${uri}, status=${status} latency=${latency_human}\n",
+		}))
+	} else {
+		e.Use(middleware.LoggerWithConfig(middleware.DefaultLoggerConfig))
+	}
+
+	// React uses a virtual router, so we need to serve the index.html for all
```

Comment:
in theory we could spin out the dash UI to a separate microservice, like we do with Ozone (Ozone UI and Ozone Backend are separate processes, separate hostnames/ports, etc).

can keep it as-is if working fine, just mentioning in case it would simplify things

Reply:
it's kinda weird; I never build the UI in development; a point in favor of separable admin-UI would be to have one UI fan out to multiple relays